include/swift/ABI/TypeIdentity.h

```diff
@@ -0,0 +1,218 @@
+//===--- TypeIdentity.h - Identity info about imported types -----*- C++ -*-==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This header declares structures useful for working with the identity of
+// a type, especially for imported types.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_ABI_TYPEIDENTITY_H
+#define SWIFT_ABI_TYPEIDENTITY_H
+
+#include "swift/Basic/LLVM.h"
+
+namespace swift {
+template <class> class TargetTypeContextDescriptor;
+struct InProcess;
+using TypeContextDescriptor = TargetTypeContextDescriptor<InProcess>;
+
+/// The different components that can appear in a TypeImportInfo sequence.
+///
+/// The declaration order here is the canonical order for these components
+/// in the sequence; the sequence is ill-formed if they are out of order.
+///
+/// Note that the ABI name (or the ordinary formal name, which
+/// immediately precedes it), symbol namespace, and related entity
+/// name together form the identity of the TypeImportInfo.
+enum class TypeImportComponent : char {
+  ABIName = 'N',
+  SymbolNamespace = 'S',
+  RelatedEntityName = 'R',
+};
+
+namespace TypeImportSymbolNamespace {
+
+/// A value for `SymbolNamespace` which indicates that this type came
+/// from a C `typedef` that was imported as a distinct type instead
+/// of a `typealias`.  Currently this can happen because:
```

Comment:
It might be better to not use "Currently" here, since I don't think we'll remember to update the comment if we ever add another case.